File: backend/seed.py
"""Seed the LoanPilot SQLite database with realistic demo data.

Usage:
    python seed.py
"""
import asyncio
import logging
import random
from datetime import datetime, timedelta

from sqlalchemy import delete

from database import AsyncSessionLocal, init_db
from models.campaign import Campaign
from models.lead import Lead
from services import eligibility

logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")
logger = logging.getLogger("seed")

random.seed(42)

FIRST_NAMES = [
    "Rahul", "Priya", "Amit", "Sneha", "Vikram", "Anjali", "Rohan", "Neha",
    "Arjun", "Pooja", "Karthik", "Divya", "Suresh", "Meera", "Rajesh", "Kavya",
    "Manish", "Aarti", "Sanjay", "Ritu", "Deepak", "Swati", "Nikhil", "Shreya",
    "Ashok", "Lakshmi", "Varun", "Ananya", "Harsh", "Ishita",
]
LAST_NAMES = [
    "Sharma", "Verma", "Patel", "Reddy", "Kumar", "Singh", "Gupta", "Iyer",
    "Nair", "Joshi", "Mehta", "Kapoor", "Das", "Chowdhury", "Bose", "Rao",
    "Menon", "Pillai", "Desai", "Shah",
]

LOAN_TYPES = ["home", "personal", "business", "auto"]
EMPLOYMENT_TYPES = ["salaried", "self_employed", "business_owner"]
STATUSES = ["called", "not_interested", "interested", "pre_qualified", "passed_to_rm"]

TRANSCRIPT_TEMPLATE = (
    "Agent: Good afternoon, I'm calling from {bank} regarding a special "
    "{loan_type} loan offer. Do you have 2 minutes?\n"
    "Customer: Yes, tell me more.\n"
    "Agent: Are you currently exploring a {loan_type} loan?\n"
    "Customer: Actually yes, I was thinking of around {amount} rupees.\n"
    "Agent: Great. May I know your monthly income and employment type?\n"
    "Customer: I earn about {income} per month, I'm {employment}.\n"
    "Agent: Thank you. Let me check your eligibility...\n"
    "Agent: Based on your profile, you {outcome}. Our Relationship Manager "
    "will reach out shortly with exact rates and next steps.\n"
    "Customer: Sounds good, thank you.\n"
    "Agent: Thank you for your time, have a great day."
)

MOCK_SUMMARY_TEMPLATE = (
    "Customer {name} expressed interest in a ₹{amount:,.0f} {loan_type} loan. "
    "They reported monthly income of ₹{income:,.0f} and are {employment}. "
    "Outcome: {outcome_line} — {next_step}."
)


def _phone() -> str:
    return "+91" + "".join(str(random.randint(0, 9)) for _ in range(10))


def _name() -> str:
    return f"{random.choice(FIRST_NAMES)} {random.choice(LAST_NAMES)}"


def _loan_amount(loan_type: str) -> float:
    if loan_type == "home":
        return float(random.choice([2500000, 4000000, 5000000, 7500000, 10000000]))
    if loan_type == "business":
        return float(random.choice([1000000, 2000000, 3500000, 5000000]))
    if loan_type == "auto":
        return float(random.choice([400000, 600000, 900000, 1200000]))
    return float(random.choice([100000, 250000, 500000, 800000]))


def _income() -> float:
    return float(random.choice([18000, 25000, 35000, 50000, 75000, 100000, 150000, 250000]))


async def seed() -> None:
    await init_db()

    async with AsyncSessionLocal() as db:
        # Wipe existing data for a clean seed
        await db.execute(delete(Lead))
        await db.execute(delete(Campaign))
        await db.commit()

        campaigns_data = [
            {"name": "Q1 Home Loan Drive", "bank_name": "HDFC Bank"},
            {"name": "Personal Loan Campaign", "bank_name": "ICICI Bank"},
            {"name": "Auto Loan March", "bank_name": "Axis Bank"},
        ]
        campaign_objs: list[Campaign] = []
        for idx, data in enumerate(campaigns_data):
            created_at = datetime.utcnow() - timedelta(days=7 - idx)
            c = Campaign(
                name=data["name"],
                bank_name=data["bank_name"],
                total_leads=0,
                status="running",
                bolna_batch_id=f"mock_batch_{idx+1:04d}",
                created_at=created_at,
            )
            db.add(c)
            campaign_objs.append(c)

        await db.commit()
        for c in campaign_objs:
            await db.refresh(c)

        # 50 leads
        total_leads = 50
        transcripts_to_fill = set(random.sample(range(total_leads), 20))

        counts: dict[int, dict[str, int]] = {
            c.id: {"called": 0, "interested": 0, "qualified": 0} for c in campaign_objs
        }

        for i in range(total_leads):
            campaign = random.choice(campaign_objs)
            # Bias loan type by campaign name
            if "Home" in campaign.name:
                loan_type = random.choices(
                    LOAN_TYPES, weights=[6, 2, 1, 1], k=1
                )[0]
            elif "Personal" in campaign.name:
                loan_type = random.choices(
                    LOAN_TYPES, weights=[1, 6, 1, 2], k=1
                )[0]
            else:
                loan_type = random.choices(
                    LOAN_TYPES, weights=[1, 2, 2, 6], k=1
                )[0]

            employment = random.choice(EMPLOYMENT_TYPES)
            loan_amount = _loan_amount(loan_type)
            monthly_income = _income()

            elig = eligibility.check(
                loan_type=loan_type,
                loan_amount=loan_amount,
                monthly_income=monthly_income,
                employment_type=employment,
            )
            eligibility_status = elig["status"]

            # Map eligibility → status but sprinkle variety
            roll = random.random()
            if roll < 0.12:
                status = "not_interested"
            elif roll < 0.22:
                status = "called"
            elif eligibility_status == "eligible":
                status = random.choice(["pre_qualified", "passed_to_rm"])
            elif eligibility_status == "review_needed":
                status = "interested"
            else:
                status = random.choice(["not_interested", "called"])

            name = _name()
            phone = _phone()

            transcript = None
            summary_text = None
            if i in transcripts_to_fill:
                outcome = (
                    "pre-qualify for this loan" if eligibility_status == "eligible"
                    else "will need additional review"
                    if eligibility_status == "review_needed"
                    else "may not qualify under current criteria"
                )
                transcript = TRANSCRIPT_TEMPLATE.format(
                    bank=campaign.bank_name or "the bank",
                    loan_type=loan_type,
                    amount=f"{loan_amount:,.0f}",
                    income=f"{monthly_income:,.0f}",
                    employment=employment.replace("_", " "),
                    outcome=outcome,
                )
                outcome_line = {
                    "eligible": "pre-qualified",
                    "review_needed": "flagged for manual review",
                    "ineligible": "did not meet criteria",
                    "pending": "pending review",
                }[eligibility_status]
                next_step = (
                    "Relationship Manager callback scheduled"
                    if eligibility_status in ("eligible", "review_needed")
                    else "lead archived"
                )
                summary_text = MOCK_SUMMARY_TEMPLATE.format(
                    name=name,
                    amount=loan_amount,
                    loan_type=loan_type,
                    income=monthly_income,
                    employment=employment.replace("_", " "),
                    outcome_line=outcome_line,
                    next_step=next_step,
                )

            created_at = datetime.utcnow() - timedelta(
                days=random.randint(0, 6),
                hours=random.randint(0, 23),
                minutes=random.randint(0, 59),
            )

            lead = Lead(
                phone=phone,
                name=name,
                campaign_id=campaign.id,
                loan_type=loan_type,
                loan_amount=loan_amount,
                monthly_income=monthly_income,
                employment_type=employment,
                status=status,
                eligibility=eligibility_status,
                call_transcript=transcript,
                summary=summary_text,
                bolna_call_id=f"bolna_call_{i+1:05d}",
                created_at=created_at,
                updated_at=created_at,
            )
            db.add(lead)

            counts[campaign.id]["called"] += 1
            if status in ("interested", "pre_qualified", "passed_to_rm"):
                counts[campaign.id]["interested"] += 1
            if status in ("pre_qualified", "passed_to_rm"):
                counts[campaign.id]["qualified"] += 1

        # Update campaign counters
        for c in campaign_objs:
            c.total_leads = counts[c.id]["called"]
            c.called_count = counts[c.id]["called"]
            c.interested_count = counts[c.id]["interested"]
            c.qualified_count = counts[c.id]["qualified"]

        await db.commit()
        logger.info(
            "Seed complete: %d campaigns, %d leads (%d with transcripts)",
            len(campaign_objs),
            total_leads,
            len(transcripts_to_fill),
        )


if __name__ == "__main__":
    asyncio.run(seed())

File: app/prisma/seed.ts
import { PrismaClient } from "../lib/generated/prisma/client";
import { PrismaPg } from "@prisma/adapter-pg";

const databaseUrl = process.env.DATABASE_URL;
if (!databaseUrl) throw new Error("DATABASE_URL is not set");

const adapter = new PrismaPg({ connectionString: databaseUrl });
const prisma = new PrismaClient({ adapter });

const NAMES = [
  "Rahul Sharma","Priya Mehta","Amit Patel","Sunita Rao","Vikram Singh",
  "Deepa Nair","Rajesh Kumar","Anita Desai","Suresh Iyer","Kavita Joshi",
  "Manish Gupta","Pooja Agarwal","Ravi Verma","Neha Chopra","Arjun Menon",
  "Shruti Pillai","Arun Mishra","Divya Krishnan","Sanjay Yadav","Meera Bose",
  "Karan Malhotra","Anjali Reddy","Vijay Nambiar","Rekha Shetty","Nikhil Jain",
  "Sonal Kapoor","Tarun Bhatt","Smita Kulkarni","Rohit Tiwari","Alka Pandey",
  "Ashok Rao","Geeta Singh","Prakash More","Usha Naik","Dinesh Patil",
  "Lata Sawant","Harish Chauhan","Ritu Thakur","Mohan Lal","Sunanda Pillai",
  "Gopal Sinha","Vandana Misra","Sunil Dubey","Radha Murthy","Pankaj Shah",
  "Jyoti Nair","Akash Tomar","Pallavi Deshpande","Yogesh Kadam","Swati Hegde",
];

const LOAN_TYPES = ["home","personal","business","auto"];
const EMPLOYMENT = ["salaried","self_employed","business_owner"];
const STATUSES = ["called","interested","pre_qualified","not_interested","passed_to_rm"];

function rand<T>(arr: T[]): T { return arr[Math.floor(Math.random() * arr.length)]; }
function randInt(min: number, max: number) { return Math.floor(Math.random() * (max - min + 1)) + min; }

const TRANSCRIPTS = [
  "Agent: Hello, this is Priya from Apex Bank. Customer: Yes, go ahead. Agent: Are you interested in a home loan? Customer: Yes, I'm looking for about 40 lakhs. Agent: What's your monthly income? Customer: Around 75,000. Agent: Are you salaried? Customer: Yes. Agent: Great, you appear eligible!",
  "Agent: Hi, calling from Apex Bank about loan offers. Customer: What kind of loans? Agent: Home, personal, business or auto. Customer: I need a personal loan of 5 lakhs. Agent: Monthly income? Customer: 35,000 salaried. Agent: You qualify for a personal loan.",
  "Agent: Good morning, Priya from Apex Bank. Customer: I'm busy, make it quick. Agent: Do you need a loan? Customer: Maybe a business loan. Agent: How much? Customer: 20 lakhs. Income around 60k, own business. Agent: Looks eligible, RM will call you.",
  "Agent: Calling about loan offers. Customer: Not interested. Agent: Understood, thank you for your time. Have a great day!",
  "Agent: Hi, this is Priya from Apex Bank. Customer: Yes? Agent: We have auto loan offers. Customer: I need a car loan, about 8 lakhs. Agent: Income? Customer: 28,000 salaried. Agent: You qualify for the auto loan!",
];

async function seed() {
  await prisma.campaign.deleteMany();
  await prisma.lead.deleteMany();

  const campaigns = await Promise.all([
    prisma.campaign.create({ data: { name: "Q1 Home Loan Drive", bankName: "Apex Bank", totalLeads: 20, calledCount: 20, interestedCount: 12, qualifiedCount: 8, status: "completed" } }),
    prisma.campaign.create({ data: { name: "Personal Loan Campaign", bankName: "Apex Bank", totalLeads: 20, calledCount: 18, interestedCount: 10, qualifiedCount: 7, status: "running" } }),
    prisma.campaign.create({ data: { name: "Auto Loan March", bankName: "Apex Bank", totalLeads: 10, calledCount: 10, interestedCount: 6, qualifiedCount: 5, status: "completed" } }),
  ]);

  for (let i = 0; i < 50; i++) {
    const name = NAMES[i];
    const phone = `+91${randInt(7000000000, 9999999999)}`;
    const loanType = rand(LOAN_TYPES);
    const employmentType = rand(EMPLOYMENT);
    const monthlyIncome = randInt(20000, 150000);
    const loanAmount = randInt(2, 80) * 100000;
    const status = rand(STATUSES);
    const campaign = campaigns[i % campaigns.length];
    const hasTranscript = i < 20;

    await prisma.lead.create({
      data: {
        phone, name, loanType, loanAmount, monthlyIncome, employmentType, status,
        eligibility: status === "pre_qualified" ? "eligible" : status === "not_interested" ? "ineligible" : "pending",
        campaignId: campaign.id,
        callTranscript: hasTranscript ? rand(TRANSCRIPTS) : null,
        summary: hasTranscript ? `${name} expressed interest in a ${loanType} loan of ₹${(loanAmount/100000).toFixed(1)}L. Monthly income ₹${monthlyIncome.toLocaleString("en-IN")}, ${employmentType}. ${status === "pre_qualified" ? "Pre-qualified — RM to follow up." : "Further review needed."}` : null,
      },
    });
  }

  console.log("✅ Seeded: 3 campaigns, 50 leads");
  await prisma.$disconnect();
}

seed().catch(console.error);
